kratos/tests/cpp_tests/utilities/test_parallel_utilities.cpp

```diff
@@ -0,0 +1,198 @@
+    }
+    class CustomReducer{
+        public:
+            double max_value = -std::numeric_limits<double>::max();
+            double max_abs = 0.0;
+
+            void LocalMerge(double function_return_value){
+                this->max_value = std::max(this->max_value,function_return_value);
+                this->max_abs   = std::max(this->max_abs,std::abs(function_return_value));
+            }
+            void ThreadSafeMerge(CustomReducer& rOther){
+                #pragma omp critical
+                {
+                this->max_value = std::max(this->max_value,rOther.max_value);
+                this->max_abs   = std::max(this->max_abs,std::abs(rOther.max_abs));
+                }
+            }
+    };
+
+    auto partition = IndexPartition<unsigned int>(data_vector.size());
+    auto ReturnValueReducer = partition.for_reduce<CustomReducer>([&](unsigned int i){
+            return data_vector[i]; //note that here the lambda returns the values to be reduced
+        });
+
+    KRATOS_CHECK_EQUAL(ReturnValueReducer.max_value, 0.0 );
+    KRATOS_CHECK_EQUAL(ReturnValueReducer.max_abs, nsize-1 );
+
+
+
+    //******************************************************************************************
+    //here we reduce at once, the sum,min,max and sub
+    typedef CombinedReduction< SumReduction<double>,
+                               MinReduction<double>,
+                               MaxReduction<double>,
+                               SubReduction<double>
+            > MultipleReduction;
+
+    auto reduction_res = IndexPartition<unsigned int>(data_vector.size()).
+        for_reduce<MultipleReduction>(
+            [&](unsigned int i){
+                    double to_sum = data_vector[i];
+                    double to_max = data_vector[i];
+                    double to_min = data_vector[i];
+                    double to_sub = data_vector[i];
+                    return std::make_tuple( to_sum, to_max, to_min, to_sub );
+                }
+            );
+    KRATOS_CHECK_EQUAL(reduction_res.GetValue<0>(), reference_sum );
+    KRATOS_CHECK_EQUAL(reduction_res.GetValue<1>(), reference_min );
+    KRATOS_CHECK_EQUAL(reduction_res.GetValue<2>(), reference_max );
+    KRATOS_CHECK_EQUAL(reduction_res.GetValue<3>(), reference_sub );
+
+
+    //check ability to handle exceptions in pure c++ - DELIBERATELY THROWING AN EXCEPTION!
+    IndexPartition<unsigned int>(data_vector.size()).for_pure_c11([&](unsigned int i){
+                    data_vector.at(i+1); //the highest thread will throw
+                }
+            );
+}
+
+KRATOS_TEST_CASE_IN_SUITE(OmpVsPureC11, KratosCoreFastSuite)
+{
+    int nsize = 1e7;
+    std::vector<double> data_vector(nsize), output(nsize);
+    for(int i=0; i<nsize; ++i)
+        data_vector[i] = i;
+
+    //check ability to handle exceptions in pure c++ - DELIBERATELY THROWING AN EXCEPTION!
+    IndexPartition<unsigned int>(data_vector.size()).for_pure_c11([&](unsigned int i){
```

Comment:
@pooyan-dadvand  this "tests" that exception is correctly captured

lines below put a small benchmark to see the relative performance of the c++11 and openmp implementation

Reply:
cool!

how is the performance?

Reply:
on my notebook (with 2 cores) 

omp time = 0.193657
pure c++11 time = 0.198034